File: tools/video-pipeline/validate_render.py
#!/usr/bin/env python3
"""Reject structurally broken final renders.

Checks stream presence, exact dimensions/orientation, A/V duration drift, and expected duration.
This catches files whose container duration looks correct while the video stream silently ends early.
"""

from __future__ import annotations

import argparse
import json
import math
import shutil
import subprocess
import sys
from pathlib import Path


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("file", type=Path)
    parser.add_argument("--orientation", choices=("portrait", "landscape", "square", "any"), default="any")
    parser.add_argument("--width", type=int)
    parser.add_argument("--height", type=int)
    parser.add_argument("--max-av-drift", type=float, default=0.25)
    parser.add_argument("--min-duration", type=float, default=0.0)
    parser.add_argument("--max-duration", type=float, default=90.0)
    args = parser.parse_args()
    if (args.width is None) != (args.height is None):
        parser.error("--width and --height must be provided together")
    if args.width is not None and (args.width <= 0 or args.height <= 0):
        parser.error("--width and --height must be positive")
    return args


def finite_float(value: object) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if math.isfinite(number) else None


def probe(path: Path) -> dict:
    if shutil.which("ffprobe") is None:
        raise SystemExit("FAIL: ffprobe is required")
    proc = subprocess.run(
        [
            "ffprobe", "-v", "error",
            "-show_entries", "format=duration:stream=index,codec_type,codec_name,width,height,duration,nb_frames",
            "-of", "json", str(path),
        ],
        check=False,
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        raise SystemExit(f"FAIL: ffprobe could not read {path}: {proc.stderr.strip()}")
    return json.loads(proc.stdout)


def main() -> int:
    args = parse_args()
    if not args.file.is_file():
        raise SystemExit(f"FAIL: file not found: {args.file}")

    data = probe(args.file)
    streams = data.get("streams") or []
    videos = [s for s in streams if s.get("codec_type") == "video"]
    audios = [s for s in streams if s.get("codec_type") == "audio"]
    if not videos:
        raise SystemExit("FAIL: no video stream")
    if not audios:
        raise SystemExit("FAIL: no audio stream")

    video, audio = videos[0], audios[0]
    width, height = int(video.get("width") or 0), int(video.get("height") or 0)
    if width <= 0 or height <= 0:
        raise SystemExit(f"FAIL: invalid dimensions {width}x{height}")

    if args.width is not None and (width != args.width or height != args.height):
        raise SystemExit(f"FAIL: expected {args.width}x{args.height}, got {width}x{height}")
    if args.orientation == "portrait" and width >= height:
        raise SystemExit(f"FAIL: expected portrait video, got {width}x{height}")
    if args.orientation == "landscape" and width <= height:
        raise SystemExit(f"FAIL: expected landscape video, got {width}x{height}")
    if args.orientation == "square" and width != height:
        raise SystemExit(f"FAIL: expected square video, got {width}x{height}")

    format_duration = finite_float((data.get("format") or {}).get("duration"))
    video_duration = finite_float(video.get("duration")) or format_duration
    audio_duration = finite_float(audio.get("duration")) or format_duration
    if video_duration is None or audio_duration is None:
        raise SystemExit("FAIL: could not determine both stream durations")

    drift = abs(video_duration - audio_duration)
    if drift > args.max_av_drift:
        raise SystemExit(
            f"FAIL: A/V durations diverge ({video_duration:.3f}s vs {audio_duration:.3f}s, drift {drift:.3f}s)"
        )

    duration = max(video_duration, audio_duration)
    if duration < args.min_duration:
        raise SystemExit(f"FAIL: render too short: {duration:.3f}s < {args.min_duration:.3f}s")
    if args.max_duration > 0 and duration > args.max_duration:
        raise SystemExit(f"FAIL: render too long: {duration:.3f}s > {args.max_duration:.3f}s")

    frames = video.get("nb_frames")
    report = {
        "file": args.file.name,
        "dimensions": f"{width}x{height}",
        "videoCodec": video.get("codec_name"),
        "audioCodec": audio.get("codec_name"),
        "videoDuration": round(video_duration, 6),
        "audioDuration": round(audio_duration, 6),
        "avDrift": round(drift, 6),
        "videoFrames": int(frames) if str(frames).isdigit() else frames,
        "status": "pass",
    }
    print(json.dumps(report, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: remotion/vrm-lipsync/scripts/validate-render.mjs
import fs from 'node:fs';
import {spawnSync} from 'node:child_process';
import path from 'node:path';

const valueArg = (name) => {
  const prefix = `--${name}=`;
  const arg = process.argv.find((item) => item.startsWith(prefix));
  return arg ? arg.slice(prefix.length) : null;
};

const inputArg = valueArg('input');
if (!inputArg) {
  console.error('使い方: node scripts/validate-render.mjs --input=out/video.mp4 [--width=720 --height=1280 --min-duration=44.5 --max-duration=45.5 --max-av-drift=0.25]');
  process.exit(2);
}

const input = path.resolve(process.cwd(), inputArg);
if (!fs.existsSync(input)) throw new Error(`render file not found: ${input}`);

const expectedWidth = valueArg('width') == null ? null : Number(valueArg('width'));
const expectedHeight = valueArg('height') == null ? null : Number(valueArg('height'));
const minDuration = Number(valueArg('min-duration') ?? 0);
const maxDuration = Number(valueArg('max-duration') ?? 0);
const maxAvDrift = Number(valueArg('max-av-drift') ?? 0.25);

if ((expectedWidth == null) !== (expectedHeight == null)) throw new Error('--width と --height は両方指定してください。');
if (expectedWidth != null && (!Number.isInteger(expectedWidth) || !Number.isInteger(expectedHeight) || expectedWidth <= 0 || expectedHeight <= 0)) {
  throw new Error('期待解像度が不正です。');
}

const probe = spawnSync('ffprobe', [
  '-v', 'error',
  '-show_entries', 'format=duration:stream=index,codec_type,codec_name,width,height,duration,nb_frames',
  '-of', 'json',
  input,
], {encoding: 'utf8'});
if (probe.error) throw probe.error;
if (probe.status !== 0) throw new Error(`ffprobe failed: ${String(probe.stderr || '').trim()}`);

const data = JSON.parse(probe.stdout || '{}');
const streams = Array.isArray(data.streams) ? data.streams : [];
const video = streams.find((stream) => stream.codec_type === 'video');
const audio = streams.find((stream) => stream.codec_type === 'audio');
if (!video) throw new Error('FAIL: no video stream');
if (!audio) throw new Error('FAIL: no audio stream');

const width = Number(video.width || 0);
const height = Number(video.height || 0);
if (!(width > 0 && height > 0)) throw new Error(`FAIL: invalid dimensions ${width}x${height}`);
if (expectedWidth != null && (width !== expectedWidth || height !== expectedHeight)) {
  throw new Error(`FAIL: expected ${expectedWidth}x${expectedHeight}, got ${width}x${height}`);
}

const formatDuration = Number(data.format?.duration);
const videoDuration = Number.isFinite(Number(video.duration)) ? Number(video.duration) : formatDuration;
const audioDuration = Number.isFinite(Number(audio.duration)) ? Number(audio.duration) : formatDuration;
if (!Number.isFinite(videoDuration) || !Number.isFinite(audioDuration)) throw new Error('FAIL: could not determine both stream durations');

const avDrift = Math.abs(videoDuration - audioDuration);
if (avDrift > maxAvDrift) {
  throw new Error(`FAIL: A/V durations diverge (${videoDuration.toFixed(3)}s vs ${audioDuration.toFixed(3)}s, drift ${avDrift.toFixed(3)}s)`);
}

const duration = Math.max(videoDuration, audioDuration);
if (Number.isFinite(minDuration) && minDuration > 0 && duration < minDuration) {
  throw new Error(`FAIL: render too short: ${duration.toFixed(3)}s < ${minDuration.toFixed(3)}s`);
}
if (Number.isFinite(maxDuration) && maxDuration > 0 && duration > maxDuration) {
  throw new Error(`FAIL: render too long: ${duration.toFixed(3)}s > ${maxDuration.toFixed(3)}s`);
}

const report = {
  file: path.basename(input),
  dimensions: `${width}x${height}`,
  videoCodec: video.codec_name || null,
  audioCodec: audio.codec_name || null,
  videoDuration: Number(videoDuration.toFixed(6)),
  audioDuration: Number(audioDuration.toFixed(6)),
  avDrift: Number(avDrift.toFixed(6)),
  videoFrames: /^\d+$/.test(String(video.nb_frames || '')) ? Number(video.nb_frames) : video.nb_frames || null,
  status: 'pass',
};
console.log(JSON.stringify(report, null, 2));
